### PR TITLE
always  Initialize user with a cart #116

### PR DESCRIPTION
**Pull request by Mikearaya**
wrong issue number used on commit message. this is for #116

### CODE REVIEW

**Review by pozylon — CHANGES_REQUESTED** [review]
Not so happy about this one:

1. it doesn't feel like mutations is the right abstraction layer to add this function, we should propably have an Orders.ensureCartForUser() function that we can call at various places more in the core of unchained. For login creating methods, the hook where we Orders.migrateCart is kind of the right place, for situations where a cart becomes an order (can happen even when a checkout function throws and also when carts get removed)
2. This could be a breaking behavior for some customers and maybe even limit some use cases, in some situations clients use "createCart" mutation to set an arbitrary orderNumber for example. So I think while it's okay to make it a default to ensure having carts on users, we should provide a "settings" parameter to disable this feature
3. We need a bootup call to ensure cart creation for all currently existing users when this feature is first activated
4. With this feature activated, we will propably have much more "dead" carts because of guest logins creating carts and that is linked to #332. To tackle that, we should remove guests after a while and make sure cart invalidation is optimized (see ticket #332 for that).

I propose to take care of 1, 2 & 3 but then kind of tag the feature "experimental" so it's only the default when some sort of experimental environment variable is set to some flag. That way we already have the feature and we can find solutions together with the needed cleanup and invalidation changes for guests and orders.

**Comment by pozylon** [comment]
@Mikearaya I started to work on a stale user removal plugin a while ago but didn't continue:

https://github.com/unchainedshop/unchained/pull/296

**Comment by Mikearaya** [comment]
@pozylon I have add the changes for 1, 2 & 3 like discussed above now there is a environment variable `ENSURE_USER_HAS_CART` required enable this feature and also a setting parameter  `assignCartForUsers` boolean value that can be passed on engine start up to enable cart creation for existing users in a system (default to false).

File: packages/api/resolvers/mutations/checkoutCart.js [hunk 1]
```diff
@@ -10,9 +10,11 @@ export default async function checkoutCart(
   log('mutation checkoutCart', { orderId, userId });
   const cart = await getCart({ orderId, user, countryContext });
   try {
-    return cart.checkout(transactionContext, {
+    const order = cart.checkout(transactionContext, {
```
Comment (pozylon) [inline packages/api/resolvers/mutations/checkoutCart.js:13]: discard please, await gone, exceptions not catched

File: packages/core-products/db/products/helpers.js [hunk 2]
```diff
@@ -119,6 +120,12 @@ Products.findProducts = ({
   const selector = {};
   const options = { sort };
 
+  log(
+    `hhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhh ${JSON.stringify(
```
Comment (pozylon) [inline packages/core-products/db/products/helpers.js:124]: hhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhh :)

File: packages/core-products/db/products/helpers.js [hunk 3]
```diff
@@ -822,3 +829,284 @@ Products.helpers({
 
 Products.getLocalizedTexts = (productId, locale) =>
   findLocalizedText(ProductTexts, { productId }, locale);
+
+const X = {
```
Comment (pozylon) [inline packages/core-products/db/products/helpers.js:833]: rm X

File: packages/core-users/db/helpers.js [hunk 4]
```diff
@@ -317,23 +326,32 @@ Users.createLoginToken = async (user, rawContext) => {
...
+      })
       .toArray();
     return (userArray || []).map((item) => new Users._transform(item)); // eslint-disable-line
   }
 
-  return Users.find(selector, { skip: offset, limit }).fetch();
+  return Users.find(buildFindSelector(query), { skip: offset, limit }).fetch();
+};
+
+Users.count = async (query) => {
```
Comment (pozylon) [inline packages/core-users/db/helpers.js:345]: Doesn't this belong to https://github.com/unchainedshop/unchained/pull/342 ?

**Review by pozylon — CHANGES_REQUESTED** [review]
The first time I have reviewed the PR, you have added `await getCart` to various mutations now this is gone and we have a dedicated Orders.ensureCartForUser function and it looks like that function gets called on checkout and on boot and it has been done in a way that is compatible with the webhooks and subscription logic, but now the function seems not to get called anymore when a login token gets created, cart gets migrated or when a cart gets removed. And then there is code from other branches and temp debugging code checked in so I think maybe you forgot to push the latest changes?

File: packages/core-products/db/products/helpers.js (hunk 2 above)
Comment (Mikearaya) [reply packages/core-products/db/products/helpers.js:124]: :disappointed: sorry i've forgot

**Comment by Mikearaya** [comment]
@pozylon you must have missed it ` Orders.ensureCartForUser` gets called in `LoginTokenCreated` handler https://github.com/unchainedshop/unchained/blob/default-user-cart/packages/platform/setup-accounts.js#L69

**Review by pozylon — CHANGES_REQUESTED** [review]
Everything looks great but we need integration tests for the ENSURE_USER_HAS_CART feature that checks the edge cases in which a user could have no cart. I'd enable ENSURE_USER_HAS_CART in jest-global-setup.js file

**Comment by pozylon** [comment]
@Mikearaya I have another request, can you extend your new function intializeUsersCart and move Orders.invalidateProviders() from core-orders/orders.js to that new function and also make it configurable, so we can solve https://github.com/unchainedshop/unchained/issues/236#issuecomment-810370355 like I suggested

**Comment by pozylon** [comment]
Going to merge this without new tests 🤟🎸 we can add those later
